### ISSUE TITLE
net: "queue: Use k_poll if enabled" commit regressed BSD Sockets performance

### ISSUE THREAD

**Issue by zephyrbot** [zephyrproject-rtos/zephyr#4022]
**_Reported by Paul Sokolovsky:_**

With a sample "http_server" written in MicroPython, with BSD Sockets as the underlying implementation, running on qemu_x86, I get:

[UPDATE: A native Zephyr testcase is available below: https://github.com/zephyrproject-rtos/zephyr/issues/4022#issuecomment-331659689]

With "git checkout 84db641de66463550fb41f083dbdf0b02fc78284"

```
$ ab -n200 http://192.0.2.1:8080/
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 192.0.2.1 (be patient)
Completed 100 requests
Completed 200 requests
Finished 200 requests


Server Software:        
Server Hostname:        192.0.2.1
Server Port:            8080

Document Path:          /
Document Length:        27 bytes

Concurrency Level:      1
Time taken for tests:   52.057 seconds
Complete requests:      200
Failed requests:        190
   (Connect: 0, Receive: 0, Length: 190, Exceptions: 0)
Total transferred:      9090 bytes
HTML transferred:       5690 bytes
Requests per second:    3.84 [#/sec] (mean)
Time per request:       260.283 [ms] (mean)
Time per request:       260.283 [ms] (mean, across all concurrent requests)
Transfer rate:          0.17 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        6  249 431.0      7    1029
Processing:     8   11   2.6     11      29
Waiting:        8   10   2.1     10      27
Total:         15  260 430.9     18    1040

Percentage of the requests served within a certain time (ms)
  50%     18
  66%     21
  75%     36
  80%   1022
  90%   1025
  95%   1026
  98%   1036
  99%   1037
 100%   1040 (longest request)
```

With "git checkout 84db641de66463550fb41f083dbdf0b02fc78284^" (note ^ at the end)

```
$ ab -n200 http://192.0.2.1:8080/
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking 192.0.2.1 (be patient)
Completed 100 requests
Completed 200 requests
Finished 200 requests


Server Software:        
Server Hostname:        192.0.2.1
Server Port:            8080

Document Path:          /
Document Length:        27 bytes

Concurrency Level:      1
Time taken for tests:   3.712 seconds
Complete requests:      200
Failed requests:        190
   (Connect: 0, Receive: 0, Length: 190, Exceptions: 0)
Total transferred:      9090 bytes
HTML transferred:       5690 bytes
Requests per second:    53.88 [#/sec] (mean)
Time per request:       18.561 [ms] (mean)
Time per request:       18.561 [ms] (mean, across all concurrent requests)
Transfer rate:          2.39 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        6    7   1.8      7      23
Processing:     8   11   2.8     11      38
Waiting:        7   10   2.4      9      35
Total:         15   19   3.3     18      44

Percentage of the requests served within a certain time (ms)
  50%     18
  66%     19
  75%     19
  80%     19
  90%     21
  95%     23
  98%     31
  99%     34
 100%     44 (longest request)
```

Sockets implementation heavily relies on FIFOs (and uses k_poll on FIFO too). ~~Per above, there's clear case of "some FIFO operation took +1s with k_fifo_get based on k_poll vs previous wait_q based impl".~~ - This was proven to not be the case (see https://github.com/zephyrproject-rtos/zephyr/issues/4022#issuecomment-333598068).

https://github.com/zephyrproject-rtos/zephyr/commit/84db641de66463550fb41f083dbdf0b02fc78284

(Imported from Jira ZEP-2593)

**Comment by pfalcon** [comment]
As an update, my further investigation so far showed that it indeed may be a (not so) usual case of butterfly effect: change in one place affects something not directly related to it. In this case, it seems that SLIP driver runs out of free RX buffers, which causes received packets to be dropped, which causes the sender (Linux host) to wait then retransmit, and that's what causes 1s delays. It's still a late-night hacking hypothesis however. Two issues in that regard however:

* There's no logging of SLIP packet dropping, so user can't have an idea where things go wrong.
* It's still kinda strange that SLIP driver started to be short of packets after exactly 84db641de66463550fb41f083dbdf0b02fc78284, so maybe that commit somehow changes the queue behavior so that packets spend longer time in the queue instead of being processed, and that's what causes shortage.

**Comment by jukkar** [comment]
> it seems that SLIP driver runs out of free RX buffers

For testing with QEMU you can increase the amount of buffers much more than with real hw. So if you want to measure some performance issues then increase the amount of buffers to several hundred or so.

**Comment by pfalcon** [comment]
> so maybe that commit somehow changes the queue behavior so that packets spend longer time in the queue instead of being processed, and that's what causes shortage.

Further investigation showed that packets don't "spend more time in queue", they're actually leaked. In a typical situation, 2 of 4 available RX net_pkt's get leaked, which causes severe performance regression.

**Comment by pfalcon** [comment]
And here's how it looks like thru the printf style debugging:

~~~
fifo 0x00401068 iface 0x0041c3e0 pkt 0x00409738 len 54
 0x00409738 ==

slip: pkt=0x004097b0
fifo 0x00401068 iface 0x0041c3e0 pkt 0x004097b0 len 54
 ==
slip: pkt=0x004097ec
thr Received pkt 0x00409738

slab_free(0x0041c310, 0x004096fc)
slab_free(0x0041c2f4, 0x00409738)
from 192.0.2.2 closed

fifo 0x00401068 iface 0x0041c3e0 pkt 0x004097ec len 74
 0x004097ec ==
thr Received pkt 0x004097ec
~~~

The above shows that at one point, pkt 0x00409738 gets added to rx_queue, and indeed, the queue contain that packet. Next pkt 0x004097b0 gets added, but after this operation, the queue is empty. Despite that, receive thread gets the original pkt 0x00409738, but not heisen-pkt 0x004097b0. Then pkt 0x004097ec is added, and gotten by the thread. 0x004097b0 is lost forever.

**Comment by pfalcon** [comment]
So, with outcomes like in the previous comment, the only reasonable explanation was that something trashes (NULLs) "head" pointer of the fifo. And not actually thrashes, but saves, NULLs, and then restores, that would be the only (still partial and ends-not-meet) explanation how put didn't work but let get still did. I did some more twists manually which didn't help (e.g. I noticed that rx_queue is preceded in memory map by a SLIP structure, subject to the suspicion of overflow, but padding it didn't help).

Then I finally went into debugger (which is in general not a panacea, because apps often behave differently under debugger). The idea was to watch rx_queue.[...].head pointer, specifically a case when head pointer is NULL, but tail pointer is non-NULL. Thru the expected random crashes and misbehavior under debugger, I arrived at sys_slist_get_not_empty() hitting my watchpoint, because it sets first head, then tail. So, I patched to do it in reverse order, and initially just refactored the expression to set tail to based on this code motion. The run under debugger produced some funky behavior, so I even thought that maybe I made a mistake during expression refactor. I went to look at sys_slist_get_not_empty() closely and saw that its code is "superfluous" - if it's run against node which is == ->tail, it means it's the only node, and the list becomes empty. So, I made the func have `list->tail = NULL`, which patch can now be seen as https://github.com/zephyrproject-rtos/zephyr/pull/4349.

The run under debugger no longer hit my watchpoint. I ran `ab -n1000` without debugger, didn't hit the issue. 1st run, 2nd, 3rd. Then I ran `ab -n10000`. I had gazillion of other debugging changes in my tree, and started to deactivate them one by one, being afraid to scare away a fix butterfly. But the situation converged - https://github.com/zephyrproject-rtos/zephyr/pull/4349 is what it took to make the issue described in this ticket non-reproducible.

So, it disappeared the same way it had appeared - with a seemingly unrelated change. Adding a being of rationality, it would mean that we either have some race condition during queue operations (shouldn't be there, we have all irq_lock protected, right?), or there's a weird issue with QEMU which exposes a race where it shouldn't be (to remind, I wasn't able to reproduce this on real HW under same conditions as on QEMU (~10s of running). A colleague looked into this and reported that he was able to "reproduce after ~30mins of running on frdm_k64f", but there're no further details, so it can be as well a different issue).
